### ISSUE TITLE
`create-sibling-gitlab` fails with error: path 'must not start or end with a special character and must not contain consecutive special characters.'

### ISSUE THREAD

**Issue by catetrai** [datalad/datalad#6681]
### What is the problem?

When creating GitLab siblings recursively using `--layout=flat` or `--layout=collection`, the creation of the subdatasets' siblings fails with the following error:

```
create_sibling_gitlab(error): code/common (dataset) [Failed to create GitLab project: GitlabCreateError(400: {'path': ['must not start or end with a special character and must not contain consecutive special characters.']})]
```

It seems that GitLab 14.9 [introduced a new regex for project name validation](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/80055) (only for newly created projects).

So the following project name formats used by DataLad in `create-sibling-gitlab` are invalid:
* `_repo_` (starts with special character)
* `{}--{}` (has consecutive special characters)

### What steps will reproduce the problem?

1. Configure a GitLab site `gitlab` in `.python-gitlab.cfg`
2. Create a dataset with nested subdatasets
3. From the superdataset directory, run `datalad create-sibling-gitlab -r --site gitlab`

### Datalad information

datalad 0.16.2

### Have you had any success using DataLad before?

Still trying it out, but very excited about it!
Also, you did a truly amazing job on the handbook 👍 
(The link to the testimonials redirects me to the main repo -- needs to be fixed?)

**Comment by bpoldrack** [comment]
Thanks for reporting (and figuring out the problem already)!

**Comment by mih** [comment]
Looks like this is a change that is conditional on a gitlab feature. The regex is

`[a-z0-9]+([._-][a-z0-9]+)*(/[a-z0-9]+([._-][a-z0-9]+)*)*`

It is not clear, which direction we should take to accommodate these changes. The current pattern is also used for gitHUB projects.

**Comment by yarikoptic** [comment]
Just FOI that original issue which triggered gitlab to add such constraints seems to root in container registry names restrictions of OCI: https://gitlab.com/gitlab-org/gitlab/-/issues/27483#note_773845513 . IMHO it is quite unfortunate that gitlab team decided to just restrict entire projects/repos naming because of the leaf use case. What would be next -- restricting because some exotic filesystem forbids some specific magical file names ( ;-) ).  Added a comment along those lines: https://gitlab.com/gitlab-org/gitlab/-/merge_requests/80055#note_1076091546

As I do not see gitlab reverting this back, we indeed should come up with an alternative. I guess we could make it gitlab specific. I do not think we would have "difficulty" of stripping special characters from both sides. The main problem is two consecutive characters.

But also -- do we really need to support `flat` hierarchy for gitlab? AFAIK it supports proper nesting and flattening was introduced primarily for github IIRC.

**Comment by catetrai** [comment]
> But also -- do we really need to support flat hierarchy for gitlab? AFAIK it supports proper nesting and flattening was introduced primarily for github IIRC.

I am in favor of keeping the `--flat` option to be able to create all datasets underneath one GitLab group. Depending on the structure / governance of the GitLab instance, one may want to avoid the proliferation of (sub)groups.

As for the path separator character, just a single `-` or `_` would do (not much of a choice there anyway, based on the regex). Using `.` would be less likely to collide with the word separator in the dataset's directory name, but I bet it's here-be-dragons territory (hidden directories? URLs? e.g. https://gitlab.com/gitlab-org/gitlab/-/issues/20184). Ideally it could be configurable on a per-dataset basis (perhaps a config property like `datalad.gitlab-SITENAME-pathsep`) so that the user can decide what makes sense for that particular dataset's naming conventions. Just my 2c!

**Comment by yarikoptic** [comment]
yeah -- we would better allow it to be configurable, it is hardcoded now at https://github.com/yarikoptic/datalad/blob/HEAD/datalad/distributed/create_sibling_gitlab.py#L410

**Comment by mih** [comment]
OK, so it sounds as if the plan could be to unify the "special chars" between `_repo_` and `some--other` to a single and identical char, and make that configurable.

That being said, it also sounds as if we need to handle more cases than the breakage we cause ourselves, because nothing would prevent a user from have forbidden names in use already.

So we probably need to introduce a dedicated validator/mangle function.

And looking at the code of the command implementation, it may be time for an overall de-spaghettification ;-)
